graylog2-server/src/main/java/org/graylog2/indexer/messages/SerializationMemoizingMessage.java

```diff
@@ -0,0 +1,186 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.ref.SoftReference;
+import java.time.Duration;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Wraps a {@link Message} by making it immutable and caching the result of {@link #serialize(SerializationContext)}
+ * calls.
+ * <p>
+ * For more details about the caching behavior, see {@link #serialize(SerializationContext)}
+ */
+public class SerializationMemoizingMessage implements ImmutableMessage {
+    private static final Logger LOG = LoggerFactory.getLogger(SerializationMemoizingMessage.class);
+    private static final Logger RATE_LIMITED_LOG = RateLimitedLogFactory.createRateLimitedLog(
+            LOG, 1, Duration.ofMinutes(1));
+
+    private final Message delegate;
+
+    private volatile SoftReference<CacheEntry> lastSerializationResult;
+
+    private record CacheEntry(byte[] serializedBytes,
+                              ObjectMapper objectMapper,
+                              Meter invalidTimeStampMeter) {}
+
+    public SerializationMemoizingMessage(Message delegate) {
+        this.delegate = delegate;
+    }
+
+    /**
+     * Serializes a message to JSON and memoizes the result as a {@link SoftReference}.
+     * <p>
+     * As long as the same {@link SerializationContext#objectMapper()} is used in consecutive calls, the result will
+     * be memoized. If a context with a different object mapper is used, the memoized serialized byte array will be
+     * replaced with a newly serialized value.
+     * <p>
+     * The serialization result will be maintained as a {@link SoftReference}. If the JVM experiences memory pressure,
+     * the memoized value might be cleared, and consecutive calls will cause the serialization operation to execute
+     * again. In that case, a warning will be logged.
+     *
+     * @param context Context required to perform the serialization
+     * @return The serialized value. May return a memoized value according to the rules described above.
+     * @throws IOException If serialization goes wrong.
+     */
+    @Override
+    public synchronized byte[] serialize(SerializationContext context)
+            throws IOException {
+
+        final var objectMapper = context.objectMapper();
+        final var invalidTimestampMeter = context.invalidTimestampMeter();
+
+        CacheEntry cachedEntry = null;
+        if (lastSerializationResult != null) {
+            cachedEntry = lastSerializationResult.get();
+            if (cachedEntry == null) {
+                RATE_LIMITED_LOG.warn("The JVM cleared a cached serialized message because of memory pressure. " +
+                        "This has a performance impact. Please adjust the memory configuration to assign more " +
+                        "heap memory to the JVM.");
+            }
+        }
+
+        if (cachedEntry == null || !cachedEntry.objectMapper().equals(objectMapper)) {
+            final var tsMeter = new Meter();
```

Comment:
The only thing that's left for me, is perhaps getting rid of the empty Meter that's appears to be unused.
But I'm fine with proceeding and looking at the actual memory impact this has when under load.